File: app/src/main/java/com/alisasadkovska/passport/ui/SettingsActivity.java
package com.alisasadkovska.passport.ui;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.alisasadkovska.passport.common.Common;
import com.alisasadkovska.passport.common.TinyDB;
import com.alisasadkovska.passport.common.Utils;
import com.alisasadkovska.passport.R;
import com.mahfa.dnswitch.DayNightSwitch;

import es.dmoral.toasty.Toasty;
import in.galaxyofandroid.spinerdialog.SpinnerDialog;
import io.github.inflationx.calligraphy3.CalligraphyConfig;
import io.github.inflationx.calligraphy3.CalligraphyInterceptor;
import io.github.inflationx.viewpump.ViewPump;
import io.github.inflationx.viewpump.ViewPumpContextWrapper;
import io.paperdb.Paper;

public class SettingsActivity extends AppCompatActivity {

    Toolbar toolbar;
    TinyDB tinyDB;
    LinearLayout view;

    TextView day_nightTxt;
    DayNightSwitch dayNightSwitch;
    Button changeBtn;

    private SpinnerDialog spinnerDialog;

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(ViewPumpContextWrapper.wrap(newBase));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ViewPump.init(ViewPump.builder()
                .addInterceptor(new CalligraphyInterceptor(
                        new CalligraphyConfig.Builder()
                                .setDefaultFontPath("fonts/NanumGothic-Regular.ttf")
                                .setFontAttrId(R.attr.fontPath)
                                .build())).build());

        tinyDB = new TinyDB(this);
        Utils.onActivityCreateSetTheme(this, tinyDB.getInt(Common.THEME_ID));
        setContentView(R.layout.activity_settings);

        toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(getString(R.string.action_settings));
        setSupportActionBar(toolbar);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
        }

        view = findViewById(R.id.view);

        changeBtn = findViewById(R.id.changeCountryBtn);
        changeBtn.setOnClickListener(view -> {
            if (Common.isConnectedToInternet(getApplicationContext())){
                showSpinner();
                spinnerDialog.showSpinerDialog();
            }

            else {
                Toasty.warning(getApplicationContext(), getString(R.string.no_internet), Toast.LENGTH_SHORT, true).show();
            }
        });

        day_nightTxt = findViewById(R.id.day_night_switchTxt);
        dayNightSwitch = findViewById(R.id.day_night_switch);
        dayNightSwitch.setDuration(500);

        if (tinyDB.getBoolean(Common.IS_DARK_MODE, true)) {
            darkModeOn();
            dayNightSwitch.setIsNight(true);
        } else
            darkModeOff();


        dayNightSwitch.setListener(isNight -> {
            if (!isNight) {
                Toasty.info(getBaseContext(), getString(R.string.dark_mode_off), Toast.LENGTH_SHORT, true).show();
                darkModeOff();
            } else {
                Toasty.info(getBaseContext(), getString(R.string.dark_mode_on), Toast.LENGTH_SHORT, true).show();
                darkModeOn();
            }
        });
    }

    private void showSpinner() {
        spinnerDialog=new SpinnerDialog(this, Paper.book().read(Common.CountryList), getString(R.string.select_your_country), R.style.DialogAnimations_SmileWindow,getString(R.string.close));

        spinnerDialog.setCancellable(true);
        spinnerDialog.setShowKeyboard(false);

        spinnerDialog.setTitleColor(getResources().getColor(R.color.colorAccent));
        spinnerDialog.setTitleColor(getResources().getColor(R.color.colorPrimaryText));
        spinnerDialog.setCloseColor(getResources().getColor(R.color.visa_required));

        spinnerDialog.bindOnSpinerListener((s, pos) -> {
            String countryName = Common.countryModel.get(pos).getName();

            toolbar.setTitle(countryName);

            Paper.book().write(Common.CountryName, countryName);
            Paper.book().write(Common.Cover, Common.countryModel.get(pos).getCover());
            Paper.book().write(Common.Latitude, Common.countryModel.get(pos).getLatitude());
            Paper.book().write(Common.Longitude, Common.countryModel.get(pos).getLongitude());

            Toasty.success(this, "changed to " + countryName, Toasty.LENGTH_SHORT).show();
        });
    }

    private void darkModeOff() {
        view.setBackgroundColor(getResources().getColor(R.color.colorWhite));
        changeBtn.setBackgroundColor(getResources().getColor(R.color.colorAccent));
        day_nightTxt.setTextColor(getResources().getColor(R.color.colorBlack));
        day_nightTxt.setText(getString(R.string.dark_mode_off));
        tinyDB.putBoolean(Common.IS_DARK_MODE, false);
        tinyDB.putInt(Common.THEME_ID,0);
    }

    private void darkModeOn() {
        view.setBackgroundColor(getResources().getColor(R.color.colorBlack));
        changeBtn.setBackgroundColor(getResources().getColor(R.color.colorAccentDark));
        day_nightTxt.setTextColor(getResources().getColor(R.color.colorWhite));
        day_nightTxt.setText(getString(R.string.dark_mode_on));
        tinyDB.putBoolean(Common.IS_DARK_MODE, true);
        tinyDB.putInt(Common.THEME_ID,1);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        if (item.getItemId() == android.R.id.home) {
            goToHome();
        }

        return super.onOptionsItemSelected(item);
    }

    private void goToHome() {
        Intent intent = new Intent(SettingsActivity.this, Home.class);
        startActivity(intent);
    }

    @Override
    public void onBackPressed() {
        goToHome();
    }
}

File: app/src/main/java/com/alisasadkovska/passport/ui/SettingsActivity.kt
package com.alisasadkovska.passport.ui

import `in`.galaxyofandroid.spinerdialog.SpinnerDialog
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.Toast
import com.alisasadkovska.passport.Model.CountryModel
import com.alisasadkovska.passport.R
import com.alisasadkovska.passport.common.BaseActivity
import com.alisasadkovska.passport.common.Common
import com.alisasadkovska.passport.common.Common.THEME_ID
import com.alisasadkovska.passport.common.TinyDB
import es.dmoral.toasty.Toasty
import io.paperdb.Paper
import kotlinx.android.synthetic.main.activity_settings.*
import java.util.*

class SettingsActivity : BaseActivity() {

    var changed = false


    private var country = ArrayList<CountryModel>()

    private lateinit var tinyDB: TinyDB
    private lateinit var spinnerDialog:SpinnerDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        tinyDB = TinyDB(this)

        toolbar.title = getString(R.string.action_settings)
        setSupportActionBar(toolbar)

        supportActionBar!!.setDisplayShowTitleEnabled(true)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)

        if (Paper.book().contains(Common.CountryModel)) {
            country = Paper.book().read(Common.CountryModel)
        }

        changeCountryBtn.setOnClickListener {
            if (Common.isConnectedToInternet(this)) {
                showSpinner()
                spinnerDialog.showSpinerDialog()
            } else {
                Toasty.warning(applicationContext, getString(R.string.no_internet), Toast.LENGTH_SHORT, true).show()
            }
        }

        day_night_switch.setDuration(500)

        when(themeId){
            0->{
                day_night_switch.setIsNight(false)
                day_night_switchTxt.text = getString(R.string.dark_mode_off)
            }
            1->{
                day_night_switch.setIsNight(true)
                day_night_switchTxt.text = getString(R.string.dark_mode_on)
            }
        }

        day_night_switch.setListener { isNight: Boolean ->
            if (isNight){
                Toasty.info(baseContext, getString(R.string.dark_mode_on), Toasty.LENGTH_SHORT).show()
                darkModeOn()
            } else {
                Toasty.info(baseContext, getString(R.string.dark_mode_off), Toasty.LENGTH_SHORT).show()
                darkModeOff()
            }
            changed = true
        }
    }

    private fun showSpinner() {
        spinnerDialog = SpinnerDialog(this, Paper.book().read<ArrayList<String>>(Common.CountryList), getString(R.string.select_your_country), getString(R.string.close))

        spinnerDialog.setCancellable(false)
        spinnerDialog.setShowKeyboard(false)

        spinnerDialog.setTitleColor(resources.getColor(R.color.darkColorPrimaryDark))
        spinnerDialog.setCloseColor(resources.getColor(R.color.visa_required))
        spinnerDialog.setItemColor(resources.getColor(R.color.visa_on_arrival_table))

        spinnerDialog.bindOnSpinerListener { _: String?, pos: Int ->
            val countryName = country[pos].name
            toolbar.title = countryName
            Paper.book().write(Common.CountryName, countryName)
            Paper.book().write(Common.Cover, country[pos].cover)
            Paper.book().write(Common.Latitude, country[pos].latitude)
            Paper.book().write(Common.Longitude, country[pos].longitude)
            Toasty.success(this, "changed to $countryName", Toasty.LENGTH_SHORT).show()
            changed = true
        }
    }



    private fun darkModeOff() {
        view.setBackgroundColor(resources.getColor(R.color.colorWhite))
        window.statusBarColor = resources.getColor(R.color.colorPrimaryDark)
        changeCountryBtn!!.setBackgroundColor(resources.getColor(R.color.colorAccent))
        day_night_switchTxt!!.setTextColor(resources.getColor(R.color.colorBlack))
        day_night_switchTxt!!.text = getString(R.string.dark_mode_off)
        toolbar.setBackgroundColor(resources.getColor(R.color.colorPrimary))
        toolbar.setTitleTextColor(resources.getColor(android.R.color.primary_text_light))
        supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_arrow_back_black_24dp)
        tinyDB.putInt(THEME_ID, 0)
    }

    private fun darkModeOn() {
        view.setBackgroundColor(resources.getColor(R.color.backgroundDark))
        window.statusBarColor = resources.getColor(R.color.darkColorPrimaryDark)
        changeCountryBtn!!.setBackgroundColor(resources.getColor(R.color.darkColorAccent))
        day_night_switchTxt!!.setTextColor(resources.getColor(R.color.colorWhite))
        day_night_switchTxt!!.text = getString(R.string.dark_mode_on)
        toolbar.setBackgroundColor(resources.getColor(R.color.darkColorPrimary))
        toolbar.setTitleTextColor(resources.getColor(android.R.color.primary_text_dark))
        supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp)
        tinyDB.putInt(THEME_ID, 1)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            goToHome()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun goToHome() {
        when {
            changed -> {
                val intent = Intent(this@SettingsActivity, HomeActivity::class.java)
                startActivity(intent)
                finish()
            }
            else -> finish()
        }


    }

    override fun onBackPressed() {
        goToHome()
    }
}